File: php/test/ConvertDirectTest.php
<?php
declare(strict_types=1);

// Convert direct test

require_once __DIR__ . '/../foreignexchangerates_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;

class ConvertDirectTest extends TestCase
{
    public function test_direct_list_convert(): void
    {
        $setup = convert_direct_setup([
            ["id" => "direct01"],
            ["id" => "direct02"],
        ]);
        [$_shouldSkip, $_reason] = Runner::is_control_skipped("direct", "direct-list-convert", $setup["live"] ? "live" : "unit");
        if ($_shouldSkip) {
            $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
            return;
        }
        if ($setup["live"]) {
            foreach (["amount01", "from01", "to01"] as $_liveKey) {
                if (!isset($setup["idmap"][$_liveKey]) || $setup["idmap"][$_liveKey] === null) {
                    $this->markTestSkipped("live test needs $_liveKey via *_ENTID env var (synthetic IDs only)");
                    return;
                }
            }
        }
        $client = $setup["client"];

        $params = [];
        if ($setup["live"]) {
            $params["amount"] = $setup["idmap"]["amount01"];
        } else {
            $params["amount"] = "direct01";
        }
        if ($setup["live"]) {
            $params["from"] = $setup["idmap"]["from01"];
        } else {
            $params["from"] = "direct01";
        }
        if ($setup["live"]) {
            $params["to"] = $setup["idmap"]["to01"];
        } else {
            $params["to"] = "direct01";
        }

        $result = $client->direct([
            "path" => "v1/convert/{from}/{to}/{amount}",
            "method" => "GET",
            "params" => $params,
        ]);
        if ($setup["live"]) {
            // Live mode is lenient: synthetic IDs frequently 4xx and the
            // list-response shape varies wildly across public APIs. Skip
            // rather than fail when the call doesn't return a usable list.
            if (!empty($result["err"])) {
                $this->markTestSkipped("list call failed (likely synthetic IDs against live API): " . (string)$result["err"]);
                return;
            }
            if (empty($result["ok"])) {
                $this->markTestSkipped("list call not ok (likely synthetic IDs against live API)");
                return;
            }
            $status = Helpers::to_int($result["status"]);
            if ($status < 200 || $status >= 300) {
                $this->markTestSkipped("expected 2xx status, got " . $status);
                return;
            }
        } else {
            $this->assertArrayNotHasKey("err", $result);
            $this->assertTrue($result["ok"]);
            $this->assertEquals(200, Helpers::to_int($result["status"]));
            $this->assertIsArray($result["data"]);
            $this->assertCount(2, $result["data"]);
            $this->assertCount(1, $setup["calls"]);
        }
    }

}


function convert_direct_setup($mockres)
{
    Runner::load_env_local();

    $calls = new \ArrayObject();

    $env = Runner::env_override([
        "FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID" => [],
        "FOREIGN_EXCHANGE_RATES_TEST_LIVE" => "FALSE",
        "FOREIGN_EXCHANGE_RATES_APIKEY" => "NONE",
    ]);

    $live = $env["FOREIGN_EXCHANGE_RATES_TEST_LIVE"] === "TRUE";

    if ($live) {
        $merged_opts = [
            "apikey" => $env["FOREIGN_EXCHANGE_RATES_APIKEY"],
        ];
        $client = new ForeignExchangeRatesSDK($merged_opts);
        return [
            "client" => $client,
            "calls" => $calls,
            "live" => true,
            "idmap" => [],
        ];
    }

    $mock_fetch = function ($url, $init) use ($calls, $mockres) {
        $calls[] = ["url" => $url, "init" => $init];
        return [
            [
                "status" => 200,
                "statusText" => "OK",
                "headers" => [],
                "json" => function () use ($mockres) {
                    if ($mockres !== null) {
                        return $mockres;
                    }
                    return ["id" => "direct01"];
                },
                "body" => "mock",
            ],
            null,
        ];
    };

    $client = new ForeignExchangeRatesSDK([
        "base" => "http://localhost:8080",
        "system" => [
            "fetch" => $mock_fetch,
        ],
    ]);

    return [
        "client" => $client,
        "calls" => $calls,
        "live" => false,
        "idmap" => [],
    ];
}

File: lua/test/convert_direct_test.lua
-- Convert direct test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("foreign-exchange-rates_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

describe("ConvertDirect", function()
  it("should direct-list-convert", function()
    local setup = convert_direct_setup({
      { id = "direct01" },
      { id = "direct02" },
    })
    local _should_skip, _reason = runner.is_control_skipped("direct", "direct-list-convert", setup.live and "live" or "unit")
    if _should_skip then
      pending(_reason or "skipped via sdk-test-control.json")
      return
    end
    if setup.live then
      for _, _live_key in ipairs({"amount01", "from01", "to01"}) do
        if setup.idmap[_live_key] == nil then
          pending("live test needs " .. _live_key .. " via *_ENTID env var (synthetic IDs only)")
          return
        end
      end
    end
    local client = setup.client

    local params = {}
    if setup.live then
      params["amount"] = setup.idmap["amount01"]
    else
      params["amount"] = "direct01"
    end
    if setup.live then
      params["from"] = setup.idmap["from01"]
    else
      params["from"] = "direct01"
    end
    if setup.live then
      params["to"] = setup.idmap["to01"]
    else
      params["to"] = "direct01"
    end

    local result, err = client:direct({
      path = "v1/convert/{from}/{to}/{amount}",
      method = "GET",
      params = params,
    })
    if setup.live then
      -- Live mode is lenient: synthetic IDs frequently 4xx and the list-
      -- response shape varies wildly across public APIs. Skip rather than
      -- fail when the call doesn't return a usable list.
      if err ~= nil then
        pending("list call failed (likely synthetic IDs against live API): " .. tostring(err))
        return
      end
      if not result["ok"] then
        pending("list call not ok (likely synthetic IDs against live API)")
        return
      end
      local status = helpers.to_int(result["status"])
      if status < 200 or status >= 300 then
        pending("expected 2xx status, got " .. tostring(status))
        return
      end
    else
      assert.is_nil(err)
      assert.is_true(result["ok"])
      assert.are.equal(200, helpers.to_int(result["status"]))
      assert.is_table(result["data"])
      assert.are.equal(2, #result["data"])
      assert.are.equal(1, #setup.calls)
    end
  end)

end)


function convert_direct_setup(mockres)
  runner.load_env_local()

  local calls = {}

  local env = runner.env_override({
    ["FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID"] = {},
    ["FOREIGN_EXCHANGE_RATES_TEST_LIVE"] = "FALSE",
    ["FOREIGN_EXCHANGE_RATES_APIKEY"] = "NONE",
  })

  local live = env["FOREIGN_EXCHANGE_RATES_TEST_LIVE"] == "TRUE"

  if live then
    local merged_opts = {
      apikey = env["FOREIGN_EXCHANGE_RATES_APIKEY"],
    }
    local client = sdk.new(merged_opts)
    return {
      client = client,
      calls = calls,
      live = true,
      idmap = {},
    }
  end

  local function mock_fetch(url, init)
    table.insert(calls, { url = url, init = init })
    return {
      status = 200,
      statusText = "OK",
      headers = {},
      json = function()
        if mockres ~= nil then
          return mockres
        end
        return { id = "direct01" }
      end,
      body = "mock",
    }, nil
  end

  local client = sdk.new({
    base = "http://localhost:8080",
    system = {
      fetch = mock_fetch,
    },
  })

  return {
    client = client,
    calls = calls,
    live = false,
    idmap = {},
  }
end
